File: Reto # - 7/Python/Extra.py
class Stack:
    def __init__(self):
        self.items = []

    def is_empty(self):
        return len(self.items) == 0

    def push(self, item):
        self.items.append(item)

    def pop(self):
        if not self.is_empty():
            return self.items.pop()
        else:
            return None

    def peek(self):
        if not self.is_empty():
            return self.items[-1]
        else:
            return None


class Queue:
    def __init__(self):
        self.items = []

    def is_empty(self):
        return len(self.items) == 0

    def enqueue(self, item):
        self.items.append(item)

    def dequeue(self):
        if not self.is_empty():
            return self.items.pop(0)
        else:
            return None


class WebBrowser:
    def __init__(self):
        self.back_stack = Stack()
        self.forward_stack = Stack()
        self.current_web_page = "Inicio"

    def navigate(self, input):
        if input == "atras":
            if not self.back_stack.is_empty():
                self.forward_stack.push(self.current_web_page)
                self.current_web_page = self.back_stack.pop()
            else:
                print("No hay páginas para retroceder.")
        elif input == "adelante":
            if not self.forward_stack.is_empty():
                self.back_stack.push(self.current_web_page)
                self.current_web_page = self.forward_stack.pop()
            else:
                print("No hay páginas para avanzar.")
        else:
            self.back_stack.push(self.current_web_page)
            self.current_web_page = input
            self.forward_stack = Stack()  # Limpiar historial adelante
        print("Página actual:", self.current_web_page)


class Printer:
    def __init__(self):
        self.print_queue = Queue()

    def receive_document(self, document):
        self.print_queue.enqueue(document)
        print(f"Documento \"{document}\" agregado a la cola de impresión.")

    def print_document(self):
        if not self.print_queue.is_empty():
            document = self.print_queue.dequeue()
            print("Imprimiendo documento:", document)
        else:
            print("No hay documentos para imprimir.")


def main():
    browser = WebBrowser()
    printer = Printer()

    print("Bienvenido al simulador de navegador web e impresora compartida.")
    print("Puede navegar a páginas web escribiendo su nombre, o usar 'atras' o 'adelante'.")
    print("Para imprimir un documento, escriba 'imprimir' seguido del nombre del documento.")

    while True:
        input_command = input("\nIngrese un comando: ").strip().lower()

        if input_command == "salir":
            print("¡Hasta luego!")
            break
        elif input_command.startswith("imprimir "):
            document = input_command[9:]
            printer.receive_document(document)
        else:
            browser.navigate(input_command)


if __name__ == "__main__":
    main()

File: Reto # - 7/Java/Extra.java
import java.util.Scanner;

// Clase para simular un navegador web
class WebBrowser {
    private Stack backStack;
    private Stack forwardStack;
    private String currentWebPage;

    public WebBrowser() {
        backStack = new Stack(10); // Tamaño máximo de la pila para historial atrás
        forwardStack = new Stack(10); // Tamaño máximo de la pila para historial adelante
        currentWebPage = "Inicio";
    }

    public void navigate(String input) {
        if (input.equals("atras")) {
            if (!backStack.isEmpty()) {
                forwardStack.push(currentWebPage);
                currentWebPage = backStack.pop();
            } else {
                System.out.println("No hay páginas para retroceder.");
            }
        } else if (input.equals("adelante")) {
            if (!forwardStack.isEmpty()) {
                backStack.push(currentWebPage);
                currentWebPage = forwardStack.pop();
            } else {
                System.out.println("No hay páginas para avanzar.");
            }
        } else {
            backStack.push(currentWebPage);
            currentWebPage = input;
            forwardStack.clear(); // Limpiar historial adelante
        }

        System.out.println("Página actual: " + currentWebPage);
    }
}

// Clase para simular una impresora compartida
class Printer {
    private Queue printQueue;

    public Printer() {
        printQueue = new Queue(10); // Tamaño máximo de la cola para documentos a imprimir
    }

    public void receiveDocument(String document) {
        printQueue.enqueue(document);
        System.out.println("Documento \"" + document + "\" agregado a la cola de impresión.");
    }

    public void printDocument() {
        if (!printQueue.isEmpty()) {
            String document = printQueue.dequeue();
            System.out.println("Imprimiendo documento: " + document);
        } else {
            System.out.println("No hay documentos para imprimir.");
        }
    }
}

public class Extra{
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Simulación de navegador web
        WebBrowser browser = new WebBrowser();

        // Simulación de impresora compartida
        Printer printer = new Printer();

        System.out.println("Bienvenido al simulador de navegador web e impresora compartida.");
        System.out.println("Puede navegar a páginas web escribiendo su nombre, o usar 'atras' o 'adelante'.");
        System.out.println("Para imprimir un documento, escriba 'imprimir' seguido del nombre del documento.");

        while (true) {
            System.out.print("\nIngrese un comando: ");
            String input = scanner.nextLine().trim().toLowerCase();

            if (input.equals("salir")) {
                System.out.println("¡Hasta luego!");
                break;
            } else if (input.startsWith("imprimir ")) {
                String document = input.substring(9);
                printer.receiveDocument(document);
            } else {
                browser.navigate(input);
            }
        }

        scanner.close();
    }
}